subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/ServerWithExpectations.groovy

```diff
@@ -43,7 +43,7 @@ abstract class ServerWithExpectations extends ExternalResource {
         resetExpectations()
     }
 
-    protected void onFailure(Throwable failure) {
+    protected synchronized void onFailure(Throwable failure) {
```

Comment:
If this class needs to be thread-safe, we should document that (also on the base class and the utilities it uses), so the next person changing these will know.

Also, why is `resetExpectations` not protected? It accesses `failure` as well.